### PR TITLE
[KernelGen][Nvidia] Add __lshift__ operator with Triton kernel

### PR DESCRIPTION
**Pull request by bwbwzzz**
## Summary
Adds a Triton kernel for `__lshift__`. Triton kernel implementation for __lshift__.

## Testing
- Validated against PyTorch reference on device side
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_lshift__.py --level core` (NVIDIA H20)

| Configuration | Torch Latency (ms) | Gems Latency (ms) | Speedup | TFLOPS |
|---|---|---|---|---|
| [1073741824] | 1.608 | 1.509 | 1.066 | 1.423 |
| [64, 64] | 0.006 | 0.006 | 0.983 | 0.001 |
| [4096, 4096] | 0.031 | 0.031 | 1.008 | 1.087 |
| [64, 512, 512] | 0.031 | 0.031 | 1.009 | 1.085 |
| [1024, 1024, 1024] | 1.663 | 1.646 | 1.010 | 1.304 |
| [1073741824] | 3.053 | 3.052 | 1.000 | 0.704 |
| [64, 64] | 0.006 | 0.006 | 1.012 | 0.001 |
| [4096, 4096] | 0.053 | 0.053 | 1.001 | 0.629 |
| [64, 512, 512] | 0.054 | 0.053 | 1.003 | 0.628 |
| [1024, 1024, 1024] | 3.057 | 3.003 | 1.018 | 0.715 |
| **Arithmetic Mean** | — | — | **1.011** | — |

## Multi-backend Testing
| Backend | Accuracy Test | Benchmark | Speedup (mean) | Notes |
|---|---|---|---|---|
| Nvidia (H20) | PASS | PASS (10 cases, --level core) | 1.011 | Primary |
| Tianshu | PASS | PASS (30 cases) | 1.003 | — |
| Muxi | PASS | PASS (30 cases) | 1.003 | — |
| Ascend | PASS | N/A | — | — |
| Hygon | PASS | PASS (30 cases) | 0.991 | — |

## Files Changed
- `src/flag_gems/ops/__lshift__.py`: Triton kernel implementation
- `tests/test_lshift__.py`: Accuracy test
- `benchmark/test_lshift__.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Math, stage: alpha 5.3)

### CODE REVIEW

**Review by Dongxu-H — DISMISSED** [review]
## Code Review: [KernelGen][Nvidia] Add sum.Int_out operator with Triton kernel

### Summary
This PR adds `sum.Int_out` operator - a specialized sum operation for integer tensors with int output. Uses `sum.Int_out` naming convention matching PyTorch ATen. CI passes (code-style ✅, python-op ✅).

---

### 🔴 Blockers

**1. Test Does Not Test the Custom Operator**

The test calls `torch.sum()` not the custom operator:
```python
@pytest.mark.Int_out
def test_sum_int_out(shape, dtype):
    inp = torch.randint(-3, 3, shape, dtype=dtype, device=flag_gems.device)
    ref_inp = utils.to_reference(inp, True)

    ref_out = torch.sum(ref_inp, dtype=torch.int64).int()
    with flag_gems.use_gems():
        res_out = torch.sum(inp, dtype=torch.int64).int()  # ← Calls torch.sum, not custom op
```

If `sum.Int_out` is registered as a custom op, the test should explicitly call it or verify that `torch.sum(..., dtype=torch.int64).int()` dispatches to the custom kernel.

**Fix:** Verify the registration path or add explicit test:
```python
from flag_gems.ops.sum import sum_int_out

# Explicit test
with flag_gems.use_gems():
    res_out = sum_int_out(inp)
```

---

**2. operators.yaml Ordering Issue**

In `conf/operators.yaml`, `sum.Int_out` is added at the end of the `sum` entries but there may be ordering issues:
```yaml
- sum:
    ...
- sum.Int_out:
    ...
```

PyTorch ATen naming: `sum.Int_out` should come after `sum.Int` (if exists) but before `sum.SymInt` (if exists). Verify alphabetical ordering within the sum group.

---

### 🟡 Suggestions

**1. Missing dim Parameter Support**

The implementation only supports full tensor sum. PyTorch's `sum` supports:
```python
torch.sum(x, dim=1, dtype=torch.int64)
```

Current kernel signature:
```python
def sum_int_out_kernel(x, mid, N: tl.constexpr):
```

This doesn't handle dimension reduction. Is `sum.Int_out` meant to only support full tensor sum?

---

**2. Autotune Config Empty**

```python
@autotune(
    configs=[],
    key=["N"],
)
```

The autotune configs are empty. Consider adding tuned configurations for better performance:
```python
@autotune(
    configs=[
        triton.Config({"BLOCK_SIZE": 256}, num_warps=4),
        triton.Config({"BLOCK_SIZE": 512}, num_warps=8),
        ...
    ],
    key=["N"],
)
```

---

**3. Test Marker Naming**

`@pytest.mark.Int_out` is too generic. Use a more specific marker:
```python
@pytest.mark.sum_Int_out
```

---

**4. Scalar Test Implementation**

The scalar test uses `torch.tensor(scalar)`:
```python
def test_sum_int_out_scalar(scalar):
    inp = torch.tensor(scalar, device=flag_gems.device)
```

Should verify this matches PyTorch behavior for 0-d tensor sum.

---

**5. Missing Edge Cases**

Consider testing:
- **Large values:** Integer overflow behavior
- **Empty tensor:** `shape=(0,)` should return 0
- **Negative integers:** Already tested with `randint(-3, 3, ...)`
- **Different dtypes:** `torch.int8`, `torch.int16`, `torch.int32`, `torch.int64`

---

**6. Registration** ⚠️

In `src/flag_gems/__init__.py`:
```python
("sum.Int_out", sum_int_out),
```

The registration uses `sum.Int_out` as the key. Verify this matches PyTorch's dispatcher pattern for `torch.sum(..., dtype=torch.int64).int()`.

---

**7. Logger Naming** ✅

```python
logger = logging.getLogger(__name__)  # → flag_gems.ops.sum.Int_out
```

Wait - the file is `sum.Int_out.py`, but Python module names with dots can cause import issues. Is the file actually named `sum.Int_out.py`?

**Issue:** Python module names should not contain dots. If the file is named `sum.Int_out.py`, rename it to `sum_int_out.py` and adjust imports accordingly.

---

### 📊 Summary

| Aspect | Status |
|--------|--------|
| Implementation | ⚠️ Check module naming |
| Tests | ❌ Doesn't test custom op |
| Benchmark | ✅ Present |
| CI | ✅ All pass |
| Configuration | ⚠️ Verify ordering |
| Module naming | ❌ Possible issue with dot in filename |

**Action Required:**
1. **Critical:** Verify file naming - Python modules should not have dots in names
2. **Required:** Fix test to actually test the custom operator
3. **Recommended:** Add autotune configs for performance
4. **Optional:** Add dim parameter support, edge case tests

File: benchmark/test_lshift__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import pytest
+import torch
+
+from . import base, consts
+
+
+@pytest.mark.lshift__
```
Comment (tengqm) [inline benchmark/test_lshift__.py]:
```suggestion
@pytest.mark.lshift
```

File: benchmark/test_lshift__.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+def test_lshift__():
```
Comment (tengqm) [inline benchmark/test_lshift__.py]:
```suggestion
def test_lshift():
```

File: benchmark/test_lshift__.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    bench = base.BinaryPointwiseBenchmark(
+        op_name="lshift__",
```
Comment (tengqm) [inline benchmark/test_lshift__.py]:
```suggestion
        op_name="lshift",
```

File: conf/operators.yaml [hunk 4]
```diff
@@ -3855,6 +3855,17 @@ ops:
       - Math
     stages:
       - alpha: '5.3'
+  - id: lshift__
```
Comment (tengqm) [inline conf/operators.yaml]:
```suggestion
  - id: lshift
```

File: conf/operators.yaml [hunk 5]
```diff
@@ -3855,6 +3855,17 @@ ops:
...
+    description: Triton kernel implementation for __lshift__.
+    for:
+      - __lshift__
+    labels:
+      - aten
+      - KernelGen
+    kind:
+      - Math
+    stages:
+      - alpha: '5.3'
```
Comment (tengqm) [inline conf/operators.yaml]:
```suggestion
      - alpha: '5.4'
```

File: src/flag_gems/ops/__lshift__.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+import logging
+
+import flag_gems  # noqa: F401
+from flag_gems.ops.bitwise_left_shift import bitwise_left_shift
+
+logger = logging.getLogger(__name__)
+
+
+def __lshift__(self, other):
+    logger.debug("GEMS __lshift__")
```
Comment (tengqm) [inline src/flag_gems/ops/__lshift__.py]:
```suggestion
    logger.debug("GEMS LSHIFT")
```

File: tests/test_lshift__.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import pytest
+import torch
+
+import flag_gems
+
+from . import accuracy_utils as utils
+
+
+@pytest.mark.lshift__
```
Comment (tengqm) [inline tests/test_lshift__.py]:
```suggestion
@pytest.mark.lshift
```

File: tests/test_lshift__.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+@pytest.mark.parametrize("shape", utils.POINTWISE_SHAPES)
+@pytest.mark.parametrize("dtype", utils.INT_DTYPES)
+def test_lshift__(shape, dtype):
```
Comment (tengqm) [inline tests/test_lshift__.py]:
```suggestion
def test_lshift(shape, dtype):
```

File: tests/test_lshift.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import pytest
```
Comment (tengqm) [inline tests/test_lshift.py:1]: rename file to `tests/test_lshift.py`

File: benchmark/test_lshift.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import pytest
```
Comment (tengqm) [inline benchmark/test_lshift.py:1]: Rename the file to `benchmark/test_lshift.py`

Comment (bwbwzzz) [reply benchmark/test_lshift.py:1]: very thanks， have revised

**Review by w1120029931-bit — APPROVED** [review]
lgtm
